[Jia Rong] iP

jiarong15 opened the pull request:
# My Duke Bot - DukeJR

#### DukeJR accelerates your workflow by tracking your daily tasks.
> No other apps better than this - Elon Musk
###### Features Include:
* Text-Based
* Simple Commands
* ~~Good~~ Superb Speed!

###### How to get started:
1. Download DukeJR from https://github.com/jiarong15/ip - Over here! :grinning::grinning:
2. Double click to enter
3. Add all your tasks
4. You're all set!

Totally **FREE!**

###### Features:
- [x] Managing Tasks
- [x] list syntax required (any unordered or ordered list supported)
- [x] Find Item with keyword
- [ ] Reminders (Coming soon!)


Make sure to start the program by `npm start` 

Here's a snippet of the main running class written in Java! :wink::wink:
```java
public static void main(String[] args) throws FileNotFoundException {
        Duke duke = new Duke(path);
        duke.run();
}
```

Commit:
Add Gradle support

jiarong15 committed:
Level 1 completed

jiarong15 committed:
Merge remote-tracking branch 'origin/add-gradle-support'
Wish to make the current copy the stable version

jiarong15 committed:
Completed Level 2

jiarong15 committed:
Level-3

jiarong15 committed:
Level 4 done

jiarong15 committed:
Level 4 done

jiarong15 committed:
Merge remote-tracking branch 'origin/master'
update

jiarong15 committed:
Finalized Level 4

jiarong15 committed:
CommandConverter class: refactor methods to include Exceptions

jiarong15 committed:
Add delete command

jiarong15 committed:
Bug fixes, refactor parts of the code

jiarong15 committed:
Add TaskMarkingHandler and Storage class to handled reading from and writing to local file

jiarong15 committed:
Add date class to help with display of dynamic dates

jiarong15 committed:
Refactor the code base to follow the recommended scheme

jiarong15 committed:
Add tests for 2 classes

jiarong15 committed:
Create jar file

jiarong15 committed:
Add java documentation on all classes and methods

This allows future developers to understand my code

read all comments above mthods and classes

jiarong15 committed:
Indent long methods and operators

Improve readability

jiarong15 committed:
Add new find command and two exceptions for find and UI

The exceptions handles if user input do not match any tasks
and if they input in the wrong format

jiarong15 committed:
Merge branch 'branch-A-CodingStandard' into branch-A-JavaDoc

To create one common master

jiarong15 committed:
Merge branch 'branch-Level-9' into branch-A-JavaDoc

tenebrius1 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,276 @@
+                            1, instructionsArray.length));
+            String identity = instructionsArray[0];
+            if (identity.equals(TODO_FULL_IDEN)) {
+                Task ToDoTask;
```
Maybe you want to install the checkstyle into your gradle package? I ran it on your files and it came up with 187 errors. Although most of them are minor things, there are multiple instances of coding standards violation like this where the variable names are not actually camelCase. Since your codebase is actually quite large automated checkstyle can probably help reduce such errors!

tenebrius1 commented on src/main/java/duke/status/typeTask.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Enumeration to indicate task name
+ */
+public enum typeTask {
```
Remember that enums are still classes and they need to be in PascalCase!

tenebrius1 reviewed (commented):
Overall I think your iP generally works well! I like that you have thoroughly documented your work which makes it easy to see what your functions do.

Some suggestions I can give is that you could perhaps add more error detection (e.g event crashes when I type `event test /at ` or `deadline test`). Your code didn't seem to have too major coding standard violations but I seen quite a few commented out imports and multiple empty lines in your code. You can consider downloading the checkstyle plugin/setup checkstyle for gradle to iron out your code to make it look nicer!

jiarong15 commented:
Good Day,


Understood. Thanks for the feedback. Will make the necessary changes







On Mon, 30 Aug 2021 at 11:25 PM, Tan Yi Guan ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Overall I think your iP generally works well! I like that you have
> thoroughly documented your work which makes it easy to see what your
> functions do.
>
> Some suggestions I can give is that you could perhaps add more error
> detection (e.g event crashes when I type event test /at or deadline test).
> Your code didn't seem to have too major coding standard violations but I
> seen quite a few commented out imports and multiple empty lines in your
> code. You can consider downloading the checkstyle plugin/setup checkstyle
> for gradle to iron out your code to make it look nicer!
> ------------------------------
>
> In src/main/java/duke/Duke.java
> <https://github.com/nus-cs2103-AY2122S1/ip/pull/181#discussion_r698557184>
> :
>
> > +    private TaskList tasks;
> +
> +
> +    /**
> +     * Instantiates Parser, Storage, Ui and Tasklist and passes the filepath to storage class
> +     * catches both exceptions from storage class. If either 1 exception happens,
> +     * Tasklist is initialized as empty, with no tasks in it.
> +     *
> +     * @param filePath String containing the relative file path
> +     * that storage class takes in to store and read tasks
> +     */
> +    public Duke(String filePath) {
> +        parser = new Parser();
> +        ui = new Ui();
> +        storage = new Storage(filePath);
> +        try {
>
> Perhaps you can consider putting the try-catch block should be inside of
> your storage class? Seems to be more appropriate to be put there since it
> is dealing with file manipulation.
> ------------------------------
>
> In src/main/java/duke/Storage.java
> <https://github.com/nus-cs2103-AY2122S1/ip/pull/181#discussion_r698572632>
> :
>
> > +        }
> +        return this.changeStorageToTasks(stringTasks);
> +    }
> +
> +    private ArrayList<Task> changeStorageToTasks(
> +            HashMap<String, Boolean> stringTasks) throws DukeException {
> +        ArrayList<Task> finalOutputTasks = new ArrayList<>();
> +        for (Map.Entry<String, Boolean> entry : stringTasks.entrySet()) {
> +            String fullInstructions = entry.getKey();
> +            String[] instructionsArray = fullInstructions.split(" ");
> +            String crucialInstructions = String.join(" ",
> +                    Arrays.copyOfRange(instructionsArray,
> +                            1, instructionsArray.length));
> +            String identity = instructionsArray[0];
> +            if (identity.equals(TODO_FULL_IDEN)) {
> +                Task ToDoTask;
>
> Maybe you want to install the checkstyle into your gradle package? I ran
> it on your files and it came up with 187 errors. Although most of them are
> minor things, there are multiple instances of coding standards violation
> like this where the variable names are not actually camelCase. Since your
> codebase is actually quite large automated checkstyle can probably help
> reduce such errors!
> ------------------------------
>
> In src/main/java/duke/status/typeTask.java
> <https://github.com/nus-cs2103-AY2122S1/ip/pull/181#discussion_r698574607>
> :
>
> > @@ -0,0 +1,23 @@
> +package duke.status;
> +
> +/**
> + * Enumeration to indicate task name
> + */
> +public enum typeTask {
>
> Remember that enums are still classes and they need to be in PascalCase!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nus-cs2103-AY2122S1/ip/pull/181#pullrequestreview-741742064>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASCDRF7O4YD4QVF7YYT4FK3T7OPIBANCNFSM5CSZCL5A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jiarong15 committed:
Add a new exception class and refactor code to handle it

Helps in situation where the user input particularly for
event and deadline has empty date or time

jiarong15 committed:
Refactor classes to amend few coding violations by checkstyle

deetoomoo commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,276 @@
+            String fullInstructions = entry.getKey();
+            String[] instructionsArray = fullInstructions.split(" ");
+            String crucialInstructions = String.join(" ",
+                    Arrays.copyOfRange(instructionsArray,
+                            1, instructionsArray.length));
+            String identity = instructionsArray[0];
+            if (identity.equals(TODO_FULL_IDEN)) {
+                Task ToDoTask;
+                if (entry.getValue()) {
+                    ToDoTask = new ToDo(crucialInstructions).
+                            updateStatus(Status.COMPLETED.getStatus());
+                } else {
+                    ToDoTask = new ToDo(crucialInstructions);
+                }
+                finalOutputTasks.add(ToDoTask);
+            } else if (identity.equals(EVENT_FULL_IDEN)) {
+                Task EventTask;
+                if (entry.getValue()) {
+                    EventTask = new Event(crucialInstructions).
+                            updateStatus(Status.COMPLETED.getStatus());
+                } else {
+                    EventTask = new Event(crucialInstructions);
+                }
+                finalOutputTasks.add(EventTask);
+            } else if (identity.equals(DEADLINE_FULL_IDEN)) {
+                Task DeadlineTask;
+                if (entry.getValue()) {
+                    DeadlineTask = new Deadline(crucialInstructions).
+                            updateStatus(Status.COMPLETED.getStatus());
+                } else {
+                    DeadlineTask = new Deadline(crucialInstructions);
+                }
+                finalOutputTasks.add(DeadlineTask);
+            } else {
+                String errorMsg = "No such task!";
+                throw new DukeException(errorMsg);
+            }
+        }
+        return finalOutputTasks;
+    }
```
This is a really long method, perhaps consider abstracting some parts of this method or removing duplicated code?

deetoomoo commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -0,0 +1,59 @@
+        StringBuilder sb = new StringBuilder("");
+        ArrayList<Task> taskItems = tasks.getTaskList();
+        int counter = 1;
+        for (int i = 1; i <= tasks.getTaskListLength(); i++) {
+            Task task = taskItems.get(i - 1);
+            String taskString = task.toString();
+            if (taskString.contains(searchString)) {
+                if (counter == 1) {
+                    sb.append(counter + ". " + taskString);
+                } else {
+                    sb.append("\n" + counter + ". " + taskString);
+                }
+                counter++;
+            }
+        }
```
This section of code is highly nested. Can consider using a guard clause maybe?

Using the following might reduce some complexity of the code too
```
(Task task: taskItems) {
    ...
}
```

deetoomoo commented on src/main/java/duke/commands/TaskCommand.java:
```diff
@@ -0,0 +1,50 @@
+            task = new ToDo(importantInstructions);
+        } else if (identity.equals(typeTask.DEADLINE.getTask())) {
+            task = new Deadline(importantInstructions);
+        } else {
+            task = new Event(importantInstructions); 
+        }
```
Don't use the else statement when you can use an if statement. See [this](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#use-the-default-branch)

deetoomoo reviewed (commented):
Good job on the iP Jia Rong! I like how descriptive your method and variable names are! One comment I have regarding code quality is, try writing smaller methods! This can be done by actively trying to reduce duplicated code and keeping each method at a single level of abstraction. Cheers :-)

jiarong15 committed:
Add the fxml files and new folders to support gui

jiarong15 committed:
Add the classpath file

jiarong15 committed:
Add minor edit

jiarong15 committed:
Refactor long methods in Date, Storage class to smaller methods.

Increase code readability and understanding for other
developers

jiarong15 committed:
Add javadoc to methods and classes I have missed.

This allows for better readability for other programmers.

jiarong15 committed:
Merge pull request #2 from jiarong15/branch-A-CodeQuality

Refactor long methods in Date, Storage class to smaller methods.

jiarong15 committed:
Add BCD-extension

jiarong15 committed:
Add Ui picture and User guide instructions

jiarong15 committed:
Update markdown

jiarong15 committed:
Add javadoc to more classes, refactor unsatisfactory parts too.

jiarong15 committed:
Add previously lost assertions

jiarong15 committed:
Refactor package name and boolean variable

jiarong15 committed:
Update

jiarong15 committed:
Add files via upload

jiarong15 committed:
Delete Ui.PNG
